Use authenticator as second factor

lukasnetz opened the pull request:
This adds a new feature where a backend user can login using a password and an authenticator instead of only using an authenticator.

lukasnetz committed:
CARL-285 added feature second factor login

lukasnetz committed:
fixed broken pipelines in travis

lukasnetz committed:
fixed code formatting

markuspoerschke commented on Classes/Service/WebAuthnAuthenticationService.php:
```diff
@@ -40,31 +46,64 @@ class WebAuthnAuthenticationService extends AuthenticationService
+            if ($loginData['uident'] == '') {
+                $loginData['uident'] = $loginData['webauthn-uident'];
+            }
+            $isProcessed = true;
+        }
+
+        return $isProcessed;
```
You could try to use the [early return pattern](https://www.itamarweiss.com/personal/2018/02/28/return-early-pattern.html) here.

markuspoerschke commented on Classes/Service/WebAuthnAuthenticationService.php:
```diff
@@ -40,31 +46,64 @@ class WebAuthnAuthenticationService extends AuthenticationService
-        $username = (string) $user['username'];
-        $data = base64_decode($this->login['uident']);
+        if ($this->backendExtensionConfiguration['secondFactorLogin'] == '1') {
+            $this->serviceChainValue = parent::authUser($user);
```
It looks like `$this->serviceChainValue` is only used as a local variable here and like it can be a simple local variable then.

```suggestion
            $serviceChainValue = parent::authUser($user);
```

lukasnetz replied:
global variable changed to local

markuspoerschke commented on Classes/Service/WebAuthnAuthenticationService.php:
```diff
@@ -40,31 +46,64 @@ class WebAuthnAuthenticationService extends AuthenticationService
         //if backendUser does not exist authentication is failed
         if ($beUser === null) {
-            return 0;
+            $serviceChainValue = 0;
```
The variable is unused, because the value gets overwritten immediately in line `118` or in line `121`.

lukasnetz replied:
fixed with new commit

lukasnetz committed:
update for feature second factor login

lukasnetz committed:
Merge branch 'master' into feature/carl-285-second-factor-login

lukasnetz committed:
wip: new tests to get more codecoverage

lukasnetz committed:
codestyle fixes

lukasnetz committed:
wip: change in tests

lukasnetz committed:
wip: changed tests

lukasnetz committed:
Fix for failing php 7.4 travis pipeline

lukasnetz committed:
new test for class WebAuthnAuthenticationService

lukasnetz committed:
reverted composer.json

lukasnetz committed:
Merge branch 'master' into feature/carl-285-second-factor-login

lukasnetz committed:
changed readme and preparing for new release

lukasnetz committed:
fix for prettyCI

markuspoerschke commented on Classes/Service/WebAuthnAuthenticationService.php:
```diff
@@ -40,44 +43,92 @@ class WebAuthnAuthenticationService extends AuthenticationService
+            return 0;
+        }
+
+        if ($this->backendExtensionConfiguration['secondFactorLogin'] == '0') {
```
```suggestion
        if (!$this->backendExtensionConfiguration['secondFactorLogin']) {
```

lukasnetz replied:
Using your suggestion.

markuspoerschke commented on Classes/Service/WebAuthnAuthenticationService.php:
```diff
@@ -40,44 +43,92 @@ class WebAuthnAuthenticationService extends AuthenticationService
+            if ($loginData['uident'] == '') {
+                $loginData['uident'] = $loginData['webauthn-uident'];
+            } else {
+                $loginData['uident-text'] = $loginData['uident'];
```
In TYPO3 the array key is named `uident_text` (with an `_` underscore instaed of a `-` minus).

lukasnetz replied:
Oversight from me. Using underscores now. Thanks

markuspoerschke commented on Resources/Private/Language/ExtConfTemplate.xlf:
```diff
@@ -0,0 +1,14 @@
+                <source>Second factor login: Login using a password and an authenticator.</source>
+            </trans-unit>
+            <trans-unit id="login.forcedToRegisterAuth">
+                <source>Registration: Force a user to registrate his authenticator on first login.</source>
```
The translation label `login.forcedToRegisterAuth` is not used anywhere in the current code base and can therefore be removed.

lukasnetz replied:
Removed the translation label

markuspoerschke commented on Resources/Private/Language/de.ExtConfTemplate.xlf:
```diff
@@ -0,0 +1,16 @@
+        <body>
+            <trans-unit id="login.secondFactorLogin">
+                <source>Second factor login: Login using a password and an authenticator.</source>
+                <target>Anmeldung mit zweitem Faktor: Anmelden mit einem passwort und einem Authenticator.</target>
```
```suggestion
                <target>Anmeldung mit zweitem Faktor: Anmelden mit einem Passwort und einem Authenticator.</target>
```

lukasnetz replied:
Fixed the typo.

markuspoerschke commented on composer.json:
```diff
@@ -26,7 +26,8 @@
         "phpunit/phpunit": "^7",
-        "typo3/minimal": "^9.5.0"
+        "typo3/minimal": "^9.5.0",
+        "phpspec/prophecy": "^1.9"
```
```suggestion
```

Prophecy is already required by PHPUnit.

lukasnetz replied:
removed phpsepc/prophecy

markuspoerschke commented on ext_conf_template.txt:
```diff
@@ -0,0 +1,4 @@
+# cat=basic/login; type=boolean; label=LLL:EXT:cvc_webauthn/Resources/Private/Language/ExtConfTemplate.xlf:login.secondFactorLogin
+secondFactorLogin = 1
+# cat=basic/login; type=boolean; label=LLL:EXT:cvc_webauthn/Resources/Private/Language/ExtConfTemplate.xlf:login.forcedToRegisterAuth
```
The setting `forcedToRegisterAuth` is never used.

lukasnetz replied:
Removed the setting.

lukasnetz committed:
Merge branch 'master' into feature/carl-285-second-factor-login

markuspoerschke commented on ext_conf_template.txt:
```diff
@@ -0,0 +1,4 @@
+# cat=basic/login; type=boolean; label=LLL:EXT:cvc_webauthn/Resources/Private/Language/ExtConfTemplate.xlf:login.secondFactorLogin
+secondFactorLogin = 1
```
```suggestion
secondFactorLogin = 0
```

If the default value is `1`, then it is a breaking change. To keep the functionality the same before and after this change, the default value should be `0`. Because not asking for the password during login is the current behaviour in the `master` branch.

lukasnetz replied:
Makes sense. Changed it back to 0. Thanks

lukasnetz committed:
Merge branch 'master' into feature/carl-285-second-factor-login

lukasnetz committed:
changes after reviewing

lukasnetz committed:
bugfixes

lukasnetz committed:
better description of new feature

lukasnetz committed:
deleted old template

markuspoerschke committed:
Fix default login

The WebAuthn login service should abstain, if no challenge response was provided.

lukasnetz committed:
fixed missing path to template

lukasnetz committed:
update readme to better describe configuration

markuspoerschke committed:
Fix unit tests
